themes/default/content/blog/converting-bicep-to-pulumi/index.md

```diff
@@ -0,0 +1,355 @@
+
+If you are curious about the specification of the AST that represents Bicep programs, head over to [this file](https://github.com/Zaid-Ajaj/pulumi-converter-bicep/blob/master/src/Converter/BicepParser.fs) from the source code. You will find types such as `BicepProgram` and `BicepSyntax` that model almost every aspect of Bicep code.
+
+As for the AST of PCL programs, you can find the type definitions [here](https://github.com/Zaid-Ajaj/pulumi-converter-bicep/blob/master/src/Converter/PulumiTypes.fs).
+
+The F# language is great for almost everything but is especially amazing for writing language converters because of its powerful pattern matching capabilities, and the ability to write code in a functional style. Many of us working at Pulumi are big fans of F#!
+
+Here is a snippet of the transformation pipeline written in F# that takes the input Bicep program and converts it to the final PCL text
+
+```fsharp
+bicepProgram
+|> BicepProgram.dropResourceUnknowns
+|> BicepProgram.reduceScopeParameter
+|> BicepProgram.parameterizeByTenantAndSubscriptionId
+|> BicepProgram.addResourceGroupNameParameterToModules
+|> BicepProgram.parameterizeByResourceGroup
+|> Transform.bicepProgramToPulumi
+|> Transform.modifyComponentPaths rewriteComponentPath
+|> Printer.printProgram
+```
+
+You can learn more on the implementation of these functions from the source code on GitHub. I've written it in a way that is easy to follow, understand and contribute to. Most of these functions are fully [unit-tested](https://github.com/Zaid-Ajaj/pulumi-converter-bicep/blob/master/src/Tests/Program.fs) and I am working on adding more tests to cover more of the Bicep language features.
+
+## Bonus converter: ARM to Pulumi
+
+Those who are familiar with the Bicep CLI, know that it has a built-in _decompiler_ that converts ARM templates to Bicep. I thought it would be fun to build another converter that takes ARM templates to Pulumi but without writing any conversion logic for it. Instead, I would decompile the ARM template to Bicep, then use the logic from the Bicep converter to generate the final PCL code. Luckily for me, the decompiler code is embedded in the nuget package [Azure.Bicep.Core](https://www.nuget.org/packages/Azure.Bicep.Core) so I didn't to rely on the existence of the Bicep CLI to decompile the ARM template.
+
+From all of this, I created another converter: `pulumi-converter-arm` that works exactly like the Bicep converter but takes ARM templates as input. Head over to the [repository](https://github.com/Zaid-Ajaj/pulumi-converter-arm) to learn more about how to install it and how to use it.
```

Comment:
Likely need to address how this relates to https://www.pulumi.com/arm2pulumi/.

Reply:
Added a paragraph ✅ 

> Depending on how well this ARM converter does its job, we might consider deprecating [arm2pulumi](https://www.pulumi.com/arm2pulumi/) in favor of it. However we first need to test it against more templates and iron out the rough edges. It would be the ideal situation because there are virtually zero maintenance costs, any improvements to the Bicep converter would automatically get picked up here. Also improvements to the decompiler from the [Azure.Bicep.Core](https://www.nuget.org/packages/Azure.Bicep.Core) nuget package would benefit the converter as well.